packages/angular_devkit/build_angular/src/builders/web-test-runner/index.ts

```diff
@@ -0,0 +1,199 @@
+// eslint-disable-next-line import/no-duplicates
+import type * as WebTestRunner from '@web/test-runner';
+// eslint-disable-next-line import/no-duplicates
+import type { EventEmitter } from '@web/test-runner';
+import { promises as fs } from 'fs';
+import path from 'path';
+import { findTestFiles } from '../../utils/test-files';
+import { buildApplicationInternal } from '../application';
+import { OutputHashing } from '../browser-esbuild/schema';
+import { logBuilderStatusWarnings } from './builder-status-warnings';
+import { WtrBuilderOptions, normalizeOptions } from './options';
+import { Schema } from './schema';
+
+export default createBuilder(
+  async (schema: Schema, ctx: BuilderContext): Promise<BuilderOutput> => {
+    ctx.logger.warn(
+      'NOTE: The Web Test Runner builder is currently EXPERIMENTAL and not ready for production use.',
+    );
+    logBuilderStatusWarnings(schema, ctx);
+
+    const options = normalizeOptions(schema);
+    const testDir = 'dist/test-out';
+
+    // Parallelize startup work.
+    const [testFiles, wtr] = await Promise.all([
+      // Glob for files to test.
+      findTestFiles(options.include, options.exclude, ctx.workspaceRoot).then((files) =>
+        Array.from(files).map((file) => path.relative(process.cwd(), file)),
+      ),
+      // Dynamic import `@web/test-runner` because as an optional peer dep, it may not be installed.
+      import('@web/test-runner').catch(() => undefined),
+      // Clean build output path.
+      fs.rm(testDir, { recursive: true, force: true }),
+    ]);
+
+    // Assert that `@web/test-runner` is installed correctly.
+    if (!wtr) {
+      return {
+        success: false,
+        // TODO(dgp1130): Display a more accurate message for non-NPM users.
+        error:
+          'Web Test Runner is not installed, most likely you need to run `npm install @web/test-runner --save-dev` in your project.',
+      };
+    }
+
+    // Build the tests and abort on any build failure.
+    const buildOutput = await buildTests(testFiles, testDir, options, ctx);
+    if (!buildOutput.success) {
+      return buildOutput;
+    }
+
+    // Run the built tests.
+    const testOutput = await runTests(wtr, `${testDir}/browser`, options);
+
+    return testOutput;
+  },
+);
+
+/** Build all the given test files and write the result to the given output path. */
+async function buildTests(
+  testFiles: string[],
+  outputPath: string,
+  options: WtrBuilderOptions,
+  ctx: BuilderContext,
+): Promise<BuilderOutput> {
+  const jasmine = path.relative(
+    process.cwd(),
+    require.resolve('jasmine-core/lib/jasmine-core/jasmine.js'),
+  );
+  const jasmineRunner = path.relative(process.cwd(), path.join(__dirname, 'jasmine_runner.js'));
+
+  // Build tests with `application` builder, using test files as entry points.
+  // Also bundle in Jasmine and the Jasmine runner script, which need to share chunked dependencies.
+  const buildOutput = await first(
```

Comment:
This seems like an unwanted behaviour as this would cause watch not to function.

Reply:
Watch mode is out of scope for this PR, but will definitely add support for it later.